# 04 — Pattern Structure Validation

**Author:** Zeineb Turki  
**Date:** April 2026  
**Phase:** 3.5 — Structure Validation  

## Objective

The triangle detector uses the **pivot + `linregress`** approach from the
*TrianglePricePatterns* reference notebook, adapted for daily SPY data.
This notebook visually validates every single detection.

**Detection pipeline (per 20-bar window):**

1. Find swing highs / lows (±3-bar neighbourhood)  
2. `scipy.stats.linregress` on pivot points → slope, intercept, correlation *r*  
3. Require |*r*| ≥ 0.9 (pivots must align tightly on the fitted line)  
4. Standard regression intercept — lines go **through** the pivots (no envelope shift)  
5. Check convergence (range compression ≥ 5%)  
6. Classify: ascending / descending / symmetric  

**Why window = 20?** The reference notebook uses `backcandles=20` on 4-hour data. For daily
SPY, 20 bars ≈ 1 month — a compact, recognisable triangle. The previous 50-bar window
(2.5 months) produced vague, overstretched patterns.

**Scope:**
- Individual chart for **every** detected triangle and channel
- Quality metrics (containment, |r|, pivot counts) on each chart
- Summary statistics

In [1]:
import sys, os
from pathlib import Path

if 'google.colab' in str(getattr(sys, 'modules', {})) or os.path.exists('/content'):
    REPO_DIR  = '/content/regime-aware-ml-trading'
    PROJ_ROOT = os.path.join(REPO_DIR, 'regime-aware-ml-trading')
    if not os.path.isdir(PROJ_ROOT):
        os.system('git clone https://github.com/zaetae/regime-aware-ml-trading.git ' + REPO_DIR)
    else:
        os.system(f'cd {REPO_DIR} && git pull -q')
    os.system(f'{sys.executable} -m pip install -q yfinance hmmlearn scikit-learn seaborn statsmodels')

    # Download SPY data if not present (data/raw/ is git-ignored)
    _spy_path = os.path.join(PROJ_ROOT, 'data', 'raw', 'spy.csv')
    if not os.path.isfile(_spy_path):
        os.makedirs(os.path.dirname(_spy_path), exist_ok=True)
        import yfinance as yf
        import pandas as pd
        _spy = yf.download('SPY', start='2010-01-01', end='2026-01-01', auto_adjust=False)
        # Newer yfinance returns MultiIndex columns — flatten them
        if isinstance(_spy.columns, pd.MultiIndex):
            _spy.columns = _spy.columns.droplevel(1)
        _spy = _spy[['Open', 'High', 'Low', 'Close', 'Volume']]
        # Strip timezone for consistent plotting
        if _spy.index.tz is not None:
            _spy.index = _spy.index.tz_localize(None)
        _spy.index.name = 'Date'
        _spy.to_csv(_spy_path)
        print(f"Downloaded SPY data to {_spy_path}")
else:
    # Find project root by looking for src/ directory
    def _find_project_root():
        current = Path.cwd()
        for _ in range(10):  # Prevent infinite loop
            if (current / "src").is_dir():
                return current
            current = current.parent
        # Fallback: assume we're in notebooks/ subdirectory
        return Path.cwd().parent if (Path.cwd().parent / "src").is_dir() else Path.cwd()

    PROJ_ROOT = str(_find_project_root())

sys.path.insert(0, PROJ_ROOT)
os.chdir(PROJ_ROOT)

%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import seaborn as sns

from src.data.load_data import load_spy
from src.patterns.export_patterns import collect_pattern_details
from src.utils.plotting import plot_candlestick, add_trendline, add_event_marker

sns.set_style('whitegrid')
plt.rcParams['figure.dpi'] = 100

# Load data and collect pattern details
df = load_spy()
all_details = collect_pattern_details(df)

# Split by pattern family
tri_details = [d for d in all_details if 'triangle' in d['pattern_type']]
ch_details  = [d for d in all_details if 'channel' in d['pattern_type']]

print(f'Loaded {len(df)} bars: {df.index[0].date()} to {df.index[-1].date()}')
print(f'Triangles: {len(tri_details)}  |  Channels: {len(ch_details)}  |  Total: {len(all_details)}')

[*********************100%***********************]  1 of 1 completed


Downloaded SPY data to /content/regime-aware-ml-trading/regime-aware-ml-trading/data/raw/spy.csv
Loaded 4024 bars: 2010-01-04 to 2025-12-31
Triangles: 15  |  Channels: 12  |  Total: 27


## 1. Triangle Detections

The triangle detector identifies converging trendlines fitted to **swing highs** (upper)
and **swing lows** (lower) via `scipy.stats.linregress` over a **20-bar window**.

Quality gates (following the reference notebook):
- |*r*| ≥ 0.9 on both trendlines (very tight pivot alignment)
- Range compression ≥ 5% (lines must converge)
- ATR-normalised slope classification

Triangle types:
- **Ascending** — flat upper (resistance) + rising lower (support)
- **Descending** — falling upper + flat lower
- **Symmetric** — both converging (falling highs, rising lows)
- **Desc. upper test** — price approaches falling resistance within a descending triangle

Below: **every** detected triangle, one chart each.

In [2]:
FORWARD_CONTEXT = 10

def plot_detection(ax, det, df):
    """Draw candlestick + pivot markers + trendlines + event marker."""
    start = det['start_idx']
    end = min(det['end_idx'] + FORWARD_CONTEXT, len(df) - 1)
    chart_slice = df.iloc[start:end + 1]
    window_slice = df.iloc[det['start_idx']:det['end_idx']]

    # Build title with quality metrics
    title = f"{det['pattern_type']}  —  {det['event_date'].strftime('%Y-%m-%d')}"
    metrics = []
    if 'containment_ratio' in det:
        metrics.append(f"containment={det['containment_ratio']:.0%}")
    if 'r_upper' in det:
        metrics.append(f"|r|=({det['r_upper']:.2f},{det['r_lower']:.2f})")
    if 'pivot_highs' in det:
        metrics.append(f"pivots=({det['pivot_highs']},{det['pivot_lows']})")
    if metrics:
        title += '\n' + '  '.join(metrics)

    plot_candlestick(chart_slice, ax=ax, title=title)

    # --- Plot swing-high and swing-low pivot markers ---
    # (like the purple markers in the reference notebook)
    if 'swing_high_idx' in det:
        for idx, price in zip(det['swing_high_idx'], det['swing_high_prices']):
            x_pos = mdates.date2num(df.index[idx])
            ax.scatter(x_pos, price, marker='v', color='red', s=60, zorder=6,
                       edgecolors='darkred', linewidths=0.5)
        # single legend entry
        ax.scatter([], [], marker='v', color='red', s=60,
                   edgecolors='darkred', label=f'Swing highs ({det["pivot_highs"]})')

    if 'swing_low_idx' in det:
        for idx, price in zip(det['swing_low_idx'], det['swing_low_prices']):
            x_pos = mdates.date2num(df.index[idx])
            ax.scatter(x_pos, price, marker='^', color='blue', s=60, zorder=6,
                       edgecolors='darkblue', linewidths=0.5)
        ax.scatter([], [], marker='^', color='blue', s=60,
                   edgecolors='darkblue', label=f'Swing lows ({det["pivot_lows"]})')

    # --- Trendlines through the pivots ---
    upper_coeffs = [det['upper_slope'], det['upper_intercept']]
    lower_coeffs = [det['lower_slope'], det['lower_intercept']]

    add_trendline(ax, window_slice, upper_coeffs, det['window'],
                  color='red', label='Upper trend')
    add_trendline(ax, window_slice, lower_coeffs, det['window'],
                  color='blue', label='Lower trend')

    # --- Detection event marker ---
    event_price = df.loc[det['event_date'], 'Close']
    add_event_marker(ax, det['event_date'], event_price,
                     marker='D', color='orange', size=80, label='Breakout')


# --- Plot EVERY triangle detection individually ---
print(f'Total triangle detections: {len(tri_details)}')
if len(tri_details) == 0:
    print('No triangles detected.')
else:
    for i, det in enumerate(tri_details):
        fig, ax = plt.subplots(figsize=(14, 5))
        plot_detection(ax, det, df)
        ax.legend(fontsize=8, loc='upper left')
        plt.tight_layout()
        plt.show()
        print()

Total triangle detections: 15

















### Triangle Observations

- **Compact patterns:** the 20-bar window produces recognisable triangle shapes spanning ~1 month
- **Lines go through the pivots:** the standard `linregress` intercept hugs the swing points tightly (|r| ≥ 0.9)
- **Convergence is clear:** upper and lower trendlines visibly narrow from left to right
- **Breakout visible:** the orange marker sits at or beyond the formation boundary, confirming the signal
- **All three types present:** ascending (flat top + rising bottom), descending (falling top + flat bottom), symmetric (both converging)

## 2. Channel Detections

The channel detector uses the same pivot + `linregress` approach with:

- |*r*| ≥ 0.85 on both trendlines
- Slopes parallel within 15%
- R² ≥ 0.70 on pivot points
- Containment ≥ 70%
- Width between 1–6× ATR

Below: **every** detected channel.

In [3]:
# --- Plot EVERY channel detection individually ---
print(f'Total channel detections: {len(ch_details)}')
if len(ch_details) == 0:
    print('No channels detected with current quality thresholds.')
else:
    for i, det in enumerate(ch_details):
        fig, ax = plt.subplots(figsize=(14, 5))
        plot_detection(ax, det, df)
        ax.legend(fontsize=9)
        plt.tight_layout()
        plt.show()
        print()

Total channel detections: 12







/tmp/ipykernel_1032/3380514404.py:7: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`). Consider using `matplotlib.pyplot.close()`.
  fig, ax = plt.subplots(figsize=(14, 5))


### Channel Observations

- Channels require the strictest combined quality: parallel slopes, high R², high containment, AND current price near a boundary
- The few surviving detections show genuine parallel structure with price respecting both bands
- Low detection count is expected — clean channels on daily SPY are rare

In [4]:
# Side-by-side: one triangle + one channel (if both exist)
has_tri = len(tri_details) > 0
has_ch = len(ch_details) > 0

if has_tri and has_ch:
    fig, axes = plt.subplots(1, 2, figsize=(16, 5))
    zoom_tri = tri_details[len(tri_details) // 2]
    zoom_ch  = ch_details[len(ch_details) // 2]
    plot_detection(axes[0], zoom_tri, df)
    axes[0].legend(fontsize=9)
    plot_detection(axes[1], zoom_ch, df)
    axes[1].legend(fontsize=9)
    fig.suptitle('Zoomed Examples — Triangle vs Channel', fontsize=13, y=1.02)
    plt.tight_layout()
    plt.show()
elif has_tri:
    zoom_tri = tri_details[len(tri_details) // 2]
    fig, ax = plt.subplots(figsize=(14, 5))
    plot_detection(ax, zoom_tri, df)
    ax.legend(fontsize=9)
    fig.suptitle('Zoomed Example — Triangle', fontsize=13, y=1.02)
    plt.tight_layout()
    plt.show()
elif has_ch:
    zoom_ch = ch_details[len(ch_details) // 2]
    fig, ax = plt.subplots(figsize=(14, 5))
    plot_detection(ax, zoom_ch, df)
    ax.legend(fontsize=9)
    fig.suptitle('Zoomed Example — Channel', fontsize=13, y=1.02)
    plt.tight_layout()
    plt.show()
else:
    print('No detections to zoom into.')

In [5]:
# Summary statistics
if len(all_details) == 0:
    print('No detections — nothing to summarise.')
else:
    details_df = pd.DataFrame(all_details)
    details_df['year'] = details_df['event_date'].dt.year

    # --- Count by pattern type ---
    fig, ax = plt.subplots(figsize=(8, 4))
    type_counts = details_df['pattern_type'].value_counts().sort_index()
    type_counts.plot.barh(ax=ax, color='steelblue')
    ax.set_xlabel('Count')
    ax.set_title('Detections by Pattern Type')
    for i, v in enumerate(type_counts.values):
        ax.text(v + 0.1, i, str(v), va='center', fontsize=9)
    plt.tight_layout()
    plt.show()

    # --- Quality metrics table ---
    agg_cols = {
        'event_date': 'size',
        'upper_slope': 'mean',
        'lower_slope': 'mean',
    }
    if 'containment_ratio' in details_df.columns:
        agg_cols['containment_ratio'] = 'mean'
    if 'r_upper' in details_df.columns:
        agg_cols['r_upper'] = 'mean'
        agg_cols['r_lower'] = 'mean'

    summary = details_df.groupby('pattern_type').agg(**{
        'count': ('event_date', 'size'),
        'first': ('event_date', 'min'),
        'last': ('event_date', 'max'),
        'avg_upper_slope': ('upper_slope', 'mean'),
        'avg_lower_slope': ('lower_slope', 'mean'),
        **({'avg_containment': ('containment_ratio', 'mean')} if 'containment_ratio' in details_df.columns else {}),
        **({'avg_r_upper': ('r_upper', 'mean')} if 'r_upper' in details_df.columns else {}),
        **({'avg_r_lower': ('r_lower', 'mean')} if 'r_lower' in details_df.columns else {}),
    })
    summary['first'] = summary['first'].dt.strftime('%Y-%m-%d')
    summary['last']  = summary['last'].dt.strftime('%Y-%m-%d')

    print('\n=== Pattern Detection Summary ===')
    print(f'Total detections: {len(all_details)}')
    print(f'  Triangles: {len(tri_details)}')
    print(f'  Channels:  {len(ch_details)}')
    print(f'  Date range: {details_df["event_date"].min().strftime("%Y-%m-%d")} to '
          f'{details_df["event_date"].max().strftime("%Y-%m-%d")}')
    if 'containment_ratio' in details_df.columns:
        print(f'  Containment: mean={details_df["containment_ratio"].mean():.3f}, '
              f'min={details_df["containment_ratio"].min():.3f}')
    print()
    display(summary)


=== Pattern Detection Summary ===
Total detections: 27
  Triangles: 15
  Channels:  12
  Date range: 2010-06-14 to 2023-09-20
  Containment: mean=0.918, min=0.800



,count,first,last,avg_upper_slope,avg_lower_slope,avg_containment,avg_r_upper,avg_r_lower
pattern_type,,,,,,,,
ascending_triangle,2,2014-03-27,2014-07-25,0.001125,0.085386,0.900000,1.00000,0.998500
channel_down,7,2012-11-13,2016-11-04,-0.170373,-0.165885,0.979429,NaN,NaN
channel_up,5,2010-10-21,2023-08-10,0.401454,0.412338,0.956200,NaN,NaN
descending_triangle,1,2023-05-24,2023-05-24,-0.439998,-0.006668,0.800000,1.00000,1.000000
symmetric_triangle,12,2010-06-14,2023-09-20,-0.382543,0.313981,0.880000,0.97825,0.996083


---

## Conclusion & Next Steps

**Detection method:**
- Trendlines fitted to **swing highs / swing lows** (±3-bar pivots) using `linregress`
- Standard regression intercept — lines go through the pivots (tight fit)
- 20-bar window → compact, recognisable triangle shapes (~1 month on daily data)
- Quality gates: |r| ≥ 0.9, convergence ≥ 5%, ATR-normalised slope classification

**Key findings:**
- 15 triangle detections across 15 years of SPY, all with |r| ≈ 1.0
- 12 channel detections
- Each detection shows a clear, compact converging/parallel pattern with visible breakout
- The 20-bar window (vs. the previous 50-bar) dramatically improves visual clarity
- Containment ranges from 80% to 100% (mean 91.8%)
**Next step:** `05_triple_barrier_labeling.ipynb` — Apply triple-barrier labels to all detected events for ML training.